watertap/unit_models/tests/unit_test_harness.py

```diff
@@ -0,0 +1,175 @@
+                "expected ".format(
+                    num=number_variables(blk), num_test=stats["number_variables"]
+                )
+            )
+        if number_total_constraints(blk) != stats["number_total_constraints"]:
+            raise UnitValueError(
+                "The number of constraints were {num}, but {num_test} was "
+                "expected ".format(
+                    num=number_total_constraints(blk),
+                    num_test=stats["number_total_constraints"],
+                )
+            )
+        if number_unused_variables(blk) != stats["number_unused_variables"]:
+            raise UnitValueError(
+                "The number of unused variables were {num}, but {num_test} was "
+                "expected ".format(
+                    num=number_unused_variables(blk),
+                    num_test=stats["number_unused_variables"],
+                )
+            )
+
+    @pytest.mark.unit
+    def test_units_consistent(self, frame_unit):
+        assert_units_consistent(frame_unit)
+
+    @pytest.mark.unit
+    def test_dof(self, frame_unit):
+        if degrees_of_freedom(frame_unit) != 0:
+            raise UnitAttributeError(
+                "The unit has {dof} degrees of freedom when 0 is required."
+                "".format(dof=degrees_of_freedom(frame_unit))
+            )
+
+    @pytest.mark.component
+    def test_initialization(self, frame_unit):
+        blk = frame_unit
+
+        # initialize
+        try:
+            blk.initialize(solver=blk._test_objs.solver, optarg=blk._test_objs.optarg)
+        except InitializationError:
+            raise InitializationError(
+                "The unit has failed to initialize successfully. Please check the output logs for more information."
+            )
+
+    @pytest.mark.component
+    def test_unit_solutions(self, frame_unit):
+        self.configure_class()
+        blk = frame_unit
+        solutions = blk._test_objs.unit_solutions
+
+        # solve unit
+        if blk._test_objs.solver is None:
+            opt = get_solver()
+        else:
+            opt = get_solver(
+                solver=blk._test_objs.solver, options=blk._test_objs.optarg
+            )
+        results = opt.solve(blk)
+
+        # check solve
+        try:
+            assert len(list(iscale.badly_scaled_var_generator(blk, zero=1e-8))) == 0
+        except AssertionError:
+            badly_scaled_var_list = iscale.badly_scaled_var_generator(blk, zero=1e-8)
+            for x in badly_scaled_var_list:
+                print(
+                    f"{x[0].name}\t{x[0].value}\tsf: {iscale.get_scaling_factor(x[0])}"
+                )
+            raise AssertionError(
+                "Check the output logs for variables that need to be re-scaled"
+            )
+        assert_optimal_termination(results)
+
+        # check results
+
+        for var, val in solutions.items():
+            if not pytest.approx(value(var), abs=1e-08, rel=1e-03) == val:
```

Comment:
I think we should do something different here, along the lines @lbianchi-lbl suggested.

A relative value of `1e-03` is very loose, different than we use elsewhere. I have two suggestions:
1. We should register these defaults in `configure_class` and let the developer overwrite the values in `configure`
2. Let `val` be an instance of `_pytest.python_api.ApproxScalar` or anything else which implements `__eq__()` but cannot be cast to `float`. See my code examples.